### ISSUE TITLE
Proposal: declaring background scripts in a neutral way

### ISSUE THREAD

**Issue by carlosjeurissen** [hober/web-extensions#282]
> Edit by @Rob--W on 2023-02-24: The spirit of this feature request is to enable cross-browser development with a single manifest file. Chrome [opposed the initial proposal](https://github.com/w3c/webextensions/issues/282#issuecomment-1309088865) that is described below, but after [proposing an alternative solution](https://github.com/w3c/webextensions/issues/282#issuecomment-1442169628) that consists of allowing `service_worker` and `scripts` to co-exist (with the latter being ignored in Chrome), the position changed to [Chrome being supportive](https://github.com/w3c/webextensions/issues/282#issuecomment-1443396585).

Since the group currently does not agree what the future of background scripting should look like Safari and Firefox with [limited event pages](https://github.com/w3c/webextensions/issues/134), Chrome with [serviceWorkers](https://developer.chrome.com/docs/extensions/mv3/service_workers/ )

To make sure developers can define their background scripts in both situations I am proposing the following syntax:
```json
"background": {
  "scripts": ["script1.js", "script2.js"]
}
```

**Browsers only supporting serviceWorkers** would use the script directly as serviceWorker if only one script is defined. If multiple script files are listed, it will create a generated serviceWorker named `_generated_service_worker.js` just like browsers currently generate a `_generated_background_page.html`. This serviceWorker will import all scripts using [importScripts()](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts):
```js
importScripts(
  'script1.js',
  'script2.js'
);
```

**Browsers only supporting limited event pages** would generate a `_generated_background_page.html` which imports the scripts using script tags:
```html
<head></head><body>
<script src="script1.js"></script>
<script src="script2.js"></script>
</body>
```

**Browsers supporting both serviceWorkers and limited event pages** would choose based on their own preference. We can let developers set `preferred_environment` to `serviceWorker` or `page` which browser vendors can ignore as such:
```json
"background": {
  "scripts": ["script1.js", "script2.js"],
  "preferred_environment": "serviceWorker"
}
```

Why is this important? This will lead to a more optimal contract between the developer and the browser. In which the developer is able to communicate their preferred environment while allowing browsers to decide what environment extensions should be using. This also makes sure browsers can change their preferred environment over time and have their own approach towards background script handling.

This also has the added benefit of allowing browsers to support both mv2 and mv3 like Safari does right now.

### `"type": "module"`
This syntax works perfectly fine with type module as well by using imports in the serviceWorker, or the type module attribute for the limited event page.

In the case of `"type": "module"`, the `_generated_service_worker.js` would look like this:
```js
import "./script1.js";
import "./script2.js";
```

In the case of [limited event pages](https://github.com/w3c/webextensions/issues/134), we can actually also start supporting `"type": "module"` by adding it to the `_generated_background_page.html` as such:
```html
<head></head><body>
<script type="module" src="script1.js"></script>
<script type="module" src="script2.js"></script>
</body>
```

See https://github.com/w3c/webextensions/issues/289

**Comment by oliverdunk** [comment]
One thing that makes this potentially less impactful is that lots of developers use bundlers which handle this sort of thing automatically. That said - adding an extra script is currently the path recommended in https://github.com/mozilla/webextension-polyfill and I think that speaks to the benefits of keeping code snippets that work in MV2 working in MV3.

**Comment by carlosjeurissen** [comment]
@oliverdunk bundlers can be very useful indeed and also using them in many of my extension projects. The point of this request was also to align declaring background scripts across browsers independent on whether or not the browsers use serviceWorkers or event pages.

Adding the webExtension polyfill is one of the use cases. Another one is reusing script files across your options pages, popup pages, content scripts and background pages. As you mentioned I also very much support the idea of reducing friction when switching manifest versions.

**Comment by apple502j** [comment]
I do believe there should be at least a way to declare both Event Pages and Service Workers in a single manifest (the script performs browser check in case a browser implements both).

This helps people downloading builds from GitHub zipball, those developing extensions on tablets or Chromebooks (yes, they do exist), or those who can't use bundlers for some reason.

**Comment by tophf** [comment]
How would automatic import work for `"type": "module"` where importScripts is not available? You can't just do `import './script2.js'` because its scope will be isolated not shared as it's in importScripts or in standard DOM scripts.

**Comment by carlosjeurissen** [comment]
@tophf That is a good catch. Since the `"type": "module"` is a case in which developers already expect isolation, having this behaviour in background scripting in extensions seems expected. If needed, developers can share scope by setting / getting properties on `globalThis` anyway. To make this extra clear, we can add it to the documentation.

### ServiceWorker with type module
So in the case of `"type": "module"`, the `_generated_service_worker.js` would look like this:
```js
import "./script1.js";
import "./script2.js";
```

### limited event pages with type module
In the case of [limited event pages](https://github.com/w3c/webextensions/issues/134), we can actually also start supporting `"type": "module"` by adding it to the `_generated_background_page.html`, see https://github.com/w3c/webextensions/issues/289

**Comment by apple502j** [comment]
Any update on this? I understand this is a lot more complicated, but being able to use one `manifest.json` file for every browser does help a lot.

**Comment by xeenon** [comment]
Safari and WebKit now support `"type": "module"` for `_generated_background_page.html` as well as for `service_worker` (just not a generated service worker.) https://github.com/WebKit/WebKit/commit/ba3a201f4e4aff8cfa206c624bd696473ace6cf2

**Comment by dotproto** [comment]
@xeenon, just to make sure I understand, with this change is the following manifest.json valid in Safari?

```json
{
  "name": "Demo",
  "version": "0.0.1",
  "manifest_version": 3,
  "background": {
    "scripts": ["lib.js", "background.js"],
    "type": "module"
  }
}
```

**Comment by dotproto** [comment]
I recently had a chance to discuss the original proposal with other Chrome folks. At the moment we have significant reservations with using a minimal set of manifest fields that work across all browsers. We expect that this approach will lead to situations where background scripts authored for page contexts or service worker contexts will be executed in a different environment from what the author expected and, as a direct result, unexpected runtime behavior. As such, we are currently opposed to the proposed introduction of a `"preferred_environment"` key in the background object or the use of a minimal set of properties that apply to multiple different execution environments.

That said, we did not discuss (and I think it may be worth exploring) other approaches to defining a single manifest file that works across multiple environments. For example, rather than have a way for a developer to suggest an execution environment, perhaps we could have an explicit declaration of entry points per environment.

**Comment by WorldLanguages** [comment]
Any updates? Chrome should suggest an alternative proposal that they wouldn't oppose.

We at the Scratch Addons open-source extension (100,000+ users) are watching this closely. Is it reasonable that updating to MV3 will force us to use a bundler? We have been considering adding a bundler to our workflow for some time, but did not foresee that we would need to hurry because of Manifest V3.

**Comment by WorldLanguages** [comment]
Any updates?

**Comment by Rob--W** [comment]
There is significant interest from extension developers in the ability to maintain one manifest file that works across browsers. A hard blocker to that is the current reality that support for event pages and service workers are inconsistent.

Chrome:
- supports `background.service_worker`
- refuses to load if `background.scripts` is present.

Safari:
- supports `background.service_worker`.
- supports `background.scripts` if `service_worker` is not specified.

Firefox:
- supports `background.service_worker` if the feature is enabled ([but there is an unintended bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1775618#c1)).
- supports `background.scripts` if `service_worker` is not specified or the feature is disabled.

I propose that we take Safari's approach here. That means that Chrome needs to raise a warning instead of a hard error when `background.scripts` is present. If that happens, then developers can specify `scripts` and `service_worker` in one manifest. Supporting browsers will select `service_worker`, and otherwise use `scripts`.

EDIT: 2024-03-14: Per https://github.com/w3c/webextensions/issues/282#issuecomment-1998039421, Safari prefers `scripts`, `page`, then `service_worker` when they are all specified; Previously `service_worker` came first. When Firefox ships `service_worker` support, it will also prefer Event pages over service workers (https://github.com/w3c/webextensions/issues/282#issuecomment-1998053719).

**Comment by xeenon** [comment]
Safari supports `page` still as a last option, after `service_worker` and `scripts`.

**Comment by erosman** [comment]
> I propose that we take Safari's approach here. That means that Chrome needs to raise a warning instead of a hard error when `background.scripts` is present. If that happens, then developers can specify `scripts` and `service_worker` in one manifest. Supporting browsers will select `service_worker`, and otherwise use `scripts`.

In that case, a _default_ option order needs to be set.

- Chrome would default to `service_worker` and ignore unsupported (silently)
- Firefox & Safari would default to `scripts`, and if not present, to `service_worker` (and lastly `page`) and ignore unsupported (silently) (e.g. if `page` & `type` are present together)

```json
"background": {
  "service_worker": "background.js",
  "scripts": ["script1.js", "script2.js"],
  "page": "background.html",
  "type": "module"
}
```

**Comment by Rob--W** [comment]
> > I propose that we take Safari's approach here. That means that Chrome needs to raise a warning instead of a hard error when `background.scripts` is present. If that happens, then developers can specify `scripts` and `service_worker` in one manifest. Supporting browsers will select `service_worker`, and otherwise use `scripts`.
> 
> In that case, a _default_ option order needs to be set.

There is no need for customizability of a default.

The order would be service_worker > scripts > page, in the order of what is available AND what the browser supports.

The motivation behind my proposal is to enable extension developers to have a single shared manifest file. This assumes that the service worker/event page logic are equivalent and interchangeable. If an extension does really not want to have a service worker, they can omit the key.

**Comment by oliverdunk** [comment]
On the Chrome side, we're happy with @Rob--W's proposal. Specifically, making a change so that extensions are still able to load when both the `service_worker` and `scripts` fields are present. We don't plan to implement any new behaviour for `scripts` and in Chrome, it will be unsupported and ignored.

Chromium bug: https://bugs.chromium.org/p/chromium/issues/detail?id=1418934

**Comment by Sjord** [comment]
Is there any functional difference between the `service_worker` and `scripts` keys? Are scripts defined here loaded in the similar ways with similar capabilities? Would it always work to put the same script file in both keys?

With Chrome supporting `service_worker` and Firefox supporting `scripts`, this can potentially be used by extension creators to differentiate between browsers. Is this a problem? E.g.

```
"background": {
  "service_worker": "chrome-specific-code.js",
  "scripts": ["firefox-specific-code.js"],
  ...
}
```

**Comment by oliverdunk** [comment]
@Sjord They are definitely different and I think what you described with different scripts for different browsers is actually part of the reason we want to take this approach. The `service_worker` field defines a script that will be registered as a service worker, which is the only way to write MV3 extensions in Chrome. On the other hand, Firefox also supports MV3 but only supports the `scripts` field which is loaded in an event page. They are fundamentally different environments which you may want different scripts for.

**Comment by Rob--W** [comment]
> Is there any functional difference between the `service_worker` and `scripts` keys? Are scripts defined here loaded in the similar ways with similar capabilities? Would it always work to put the same script file in both keys?

The environments are different, but if your extension only uses the set of APIs common to both environments, then your extension should work. Common to both environments means: no Service worker-specific APIs and no event page-specific (DOM) APIs. If your extension consists of only extension APIs and only uses standard JavaScript features not dependent on DOM, then you can use the same script for both the `service_worker` and `scripts` key.

While in practice Firefox does *currently* not support `service_worker`, I recommend against assuming `scripts == Firefox` or `service_worker == Chrome`. It is possible for a future Firefox release to support `service_worker`, and at that point, the Service worker would be loaded instead when both are specified.

**Comment by erosman** [comment]
> The order would be service_worker > scripts > page, in the order of what is available AND what the browser supports.
> 
> The motivation behind my proposal is to enable extension developers to have a single shared manifest file. This assumes that the service worker/event page logic are equivalent and interchangeable. If an extension does really not want to have a service worker, they can omit the key.

### Purpose
Using the same `manifest.json` for all browsers

### Possibilities
There are 2 possibilities when handling background script.

**Common Background Script**
Same background script is used that works both as SW & EP (event page). This script does not use any feature specific to SW or EP; or use logic when applying them.
In this case, default order **does not** matter and `background.service_worker` is an alias for `background.script`
```json
"background": {
  "service_worker": "background.js",
  "scripts": ["background.js"],
  "type": "module"
}
```

**Specific Background Script**
As in my initial example, different scripts are used for different browsers. 
In this case, default order **does** matter.
- Chrome loads `background.js`
- Firefox/Safari load `script1.js`, `script2.js`

```json
"background": {
  "service_worker": "background.js",
  "scripts": ["script1.js", "script2.js"],
  "type": "module"
}
```

**Comment by hanguokai** [comment]
As I [commented](https://github.com/w3c/webextensions/issues/14#issuecomment-1444011993) in #12, it's better to establish an extension manifest specification first. And this issue is part of the spec.

**Comment by xeenon** [comment]
> **Specific Background Script** As in my initial example, different scripts are used for different browsers. In this case, default order **does** matter.
> 
> * Chrome loads `background.js`
> * Firefox/Safari load `script1.js`, `script2.js`

Safari supports `service_worker` and prefers it if `service_worker` and `scripts` are present. The preferred order is what @Rob--W outlined [here](https://github.com/w3c/webextensions/issues/282#issuecomment-1443540563).

**Comment by erosman** [comment]
> Safari supports `service_worker` and prefers it if `service_worker` and `scripts` are present. The preferred order is what @Rob--W outlined [here](https://github.com/w3c/webextensions/issues/282#issuecomment-1443540563).

In that case, I misunderstood. I was under the impression that Firefox/Safari implemented background Limited Event Pages as their preferred choice in comparison to background Service Worker. 

If the preferred option was/is going to be Service Worker, wouldn't that make the development of Limited Event Pages rather redundant or secondary?

**Comment by erosman** [comment]
> Safari supports `service_worker` and prefers it if `service_worker` and `scripts` are present. The preferred order is what @Rob--W outlined [here](https://github.com/w3c/webextensions/issues/282#issuecomment-1443540563).

```json
"background": {
  "service_worker": "background.js",
  "scripts": ["script1.js", "script2.js"],
  "type": "module"
}
```

Therefore, in future, if both properties are included and supported, `service_worker` will be loaded, thus making `scripts` redundant.
If the intension is to load `scripts`, then separate `manifest.json` has to be used for each browser.

**Comment by asamuzaK** [comment]
> We don't plan to implement any new behaviour for scripts and in Chrome, it will be unsupported and ignored.

If that's the case, the default order should be
1. scripts
2. service_worker
3. page

**Comment by WorldLanguages** [comment]
Is this being tracked on crbug/bugzilla? Anyone have the link?

**Comment by oliverdunk** [comment]
> Is this being tracked on crbug/bugzilla? Anyone have the link?

In Chrome we are planning to allow service_worker and scripts in the manifest simultaneously, and will ignore scripts. It's tracked here: https://bugs.chromium.org/p/chromium/issues/detail?id=1418934

**Comment by fregante** [comment]
@oliverdunk with Firefox already shipping MV3 without service worker support, it’d be great to see Chrome allow a `background.script` property sooner rather than later. 

A lot of extensions have been stuck on MV2 due to Firefox, but this would unblock many of them (without manual manifest.json intervention)

**Comment by WorldLanguages** [comment]
Any news? Looks like Chrome hasn't updated their manifest.json validator yet, so you still can't use the same manifest file in all browsers.

**Comment by oliverdunk** [comment]
> Any news? Looks like Chrome hasn't updated their manifest.json validator yet, so you still can't use the same manifest file in all browsers.

Still very much supportive of this but as you mentioned, we haven't made any changes yet. I've started work on this today, hopefully we can make the change soon.

**Comment by carlosjeurissen** [comment]
@oliverdunk very much looking forward to this. Keep in mind besides `scripts`, Chrome should also ignore `page` and `persistent` for maximum compatibility.

**Comment by WorldLanguages** [comment]
> In Chrome 120, we will have finished addressing all our prioritized platform gaps and closed all critical bugs that are documented on the [known issues page](https://developer.chrome.com/docs/extensions/migrating/known-issues/).

(Source: [What's happening in Chrome Extensions? - developer.chrome.com](https://developer.chrome.com/blog/extension-news-october-2023/))

Even though this isn't listed in the known issues page, I hope this will be fixed in Chrome before any timeline announcement, instead of after.

**Comment by carlosjeurissen** [comment]
@xeenon In case both `background.page` and `background.service_worker` are defined, would Safari be supportive of allowing extension authors to specify a preference of using a `background.page` over a `background.service_worker`? Use case being a consistent manifest across browsers, and extensions working in their ideal environment (for example, them relying heavily on document APIs).

**Comment by xeenon** [comment]
@carlosjeurissen Yes, we would be supportive of that. Page is currently better supported in Safari still.

**Comment by WorldLanguages** [comment]
As seen above, Chrome stated they plan to fix their JSON validator. It should ignore unknown keys (in the "background" object) the same way it ignores unknown permissions: a warning, not an error. It would also match the existing behavior when encountering unknown keys elsewhere in the manifest.

However, the following behavior from Firefox is making me very confused. Firefox considers `background.service_worker` an "unexpected property", but **it acts differently when encountered with it**, and for some reason the background script doesn't run! Changing it to `background.service_workerrrr` (or anything else) fixes the issue. What's going on?

![Code_MDF0nZlNZY](https://github.com/w3c/webextensions/assets/17484114/a2075fe3-a716-4e1a-a6f2-a8b6e5372a06)

Achieving cross-browser compatibility would be quite simple:
- Browser versions that support *service workers only* should completely ignore the `scripts`, `page`, `persistent` keys and/or any other unknown keys, possibly printing a warning. **Chrome is already tracking this in [crbug](https://bugs.chromium.org/p/chromium/issues/detail?id=1418934), fortunately.**
- Browser versions that support *event pages only* should **completely** ignore the `service_worker` key. Why is Firefox reacting differently to it? I believe this is mentioned in https://github.com/mozilla/web-ext/issues/2532#issuecomment-1500717723 and it obtained 70 upvotes.
- If Firefox supports both in the future, it should first check if the `service_worker` key exists. If it does, it should behave like the first item (Chrome behavior), else see the second item (current Firefox behavior). Edit: these browsers could also support a preferred environment property and Chrome would unintentionally ignore it. **Worth considering that an event page may register a service worker, but not viceversa, so choosing eventpage over sw might be better. Chrome doesn't plan to support both, so it wouldn't be affected**

Could this be prioritized? This has been the main MV3 upgrade blocker for multiple free/open-source extensions we work on in our spare time. It would be a pity to add and maintain a build step just because of JSON validator choices by each of the browsers.

**Comment by Rob--W** [comment]
> However, the following behavior from Firefox is making me very confused. Firefox considers `background.service_worker` an "unexpected property", but **it acts differently when encountered with it**, and for some reason the background script doesn't run! Changing it to `background.service_workerrrr` (or anything else) fixes the issue. What's going on?

Thanks for highlighting this issue in Firefox. I investigated this issue and filed a bug at https://bugzilla.mozilla.org/show_bug.cgi?id=1860304 and will submit a patch next week.

**Comment by erosman** [comment]
With apologies, can the following be clarified?

Consider wanting to use the same `manifest.json` for all browsers e.g.

```json
"background": {
  "service_worker": "background.js",
  "scripts": ["script.js"],
  "type": "module"
}
```

## Case 1
When a browser supports EITHER "service_worker" OR "scripts", then the browser loads the supported type.
No problem there.

## Case 2
When in future browsers supports BOTH  "service_worker" AND "scripts", all browsers will load  "service_worker" !!?
If that is the case, including "scripts" becomes meaningless and might as well be omitted.

If an extension wants to use "scripts", there is no option but to have separate `manifest.json` for each browser, which defeats the purpose of  `manifest.json` unification.

##  Idea
Browsers have already prioritized their preferred background script type by developing that type first.

Chrome -> service_worker
Firefox & Safari -> event page

Why not maintain the original intension?

That would mean:

- Chrome would always load "service_worker"
- Firefox & Safari
  1. Load  "scripts" if present
  2. Load  "service_worker" (if above not present)
  
## Result
  
 This method would be more versatile as developers can decide what they want to load in each browser.

- If they want  "service_worker" for all browser, then they will only include  "service_worker"
- If they want  "service_worker" for Chrome and  "scripts" for others, then they will include both

**Comment by Rob--W** [comment]
> With apologies, can the following be clarified?
> 
> (...)
> When in future browsers supports BOTH  "service_worker" AND "scripts", all browsers will load  "service_worker" !!?

Yes. This was proposed at https://github.com/w3c/webextensions/issues/282#issuecomment-1442169628

Note that Chrome has no intention of supporting "scripts" in MV3. They will change the implementation to treat the presence of "scripts" as a warning instead of a hard error.

> If that is the case, including "scripts" becomes meaningless and might as well be omitted.

The current outlook is that Chrome will be service worker-only, while Firefox and Safari will support both.

> If an extension wants to use "scripts", there is no option but to have separate `manifest.json` for each browser, which defeats the purpose of  `manifest.json` unification.

The main idea behind the scripts+service_worker precedence order is that extensions can work across browsers with a single manifest file. Extensions that only use extension APIs, independent of DOM (and not using the few service worker-specific APIs) will work regardless of execution environment. The scenario of preferring a DOM-based background page isn't available in Chrome, and therefore there is no point in defaulting to supporting event pages over service workers.

Similar to what Timothy mentioned above on behalf of Apple/Safari (https://github.com/w3c/webextensions/issues/282#issuecomment-1771124037), I would be supportive of including an option to specify the preferred environment in Firefox.

**Comment by WorldLanguages** [comment]
Yes, Firefox could add a "preferred environment" key in the future. It would be fine as long as Chrome doesn't hard-error when seeing that unsupported key. As far as I know, that wouldn't be a problem as Chrome only hard-errors when it finds certain keys, and they even plan to convert those to warnings.

**Comment by xeenon** [comment]
Okay, what should the "preferred environment" key be? The original proposal was `preferred_environment` set to either `serviceWorker` or `page`. I tend to prefer `service_worker` over `serviceWorker` since it matches the keys at the top level.

**Comment by erosman** [comment]
AFA "preferred environment" key, why not use the same key as in "background"?
That way, there wont be a need for mapping it  i.e.

```
{
  "preferred_environment": "service_worker" or "scripts"
}
```

**Comment by tophf** [comment]
Wouldn't it be simpler and intuitive for developers if browsers simply treat this section as ordered dictionary?

**Comment by xeenon** [comment]
There is no such thing as an ordered dictionary in JSON / JavaScript once it is parsed. We would not be able to support that in Safari / WebKit at least, since the parsed manifest is long removed from the string / data form by the point we decided what method to use.

Per the JSON spec: ["An object is an unordered set of name/value pairs."](https://www.json.org/json-en.html#:~:text=An%20object%20is%20an%20unordered%20set%20of%20name/value%20pairs.)

**Comment by dotproto** [comment]
Yeah, an array of objects is the closest thing JSON allows. It’s a bit more
verbose, but it may be preferable to having an implicit resolution order as
it keeps the developer in control and ordering explicit.

**Comment by tophf** [comment]
JSON specification doesn't really matter in this particular case as order can be preserved e.g. by using a customized parser. Seems pretty trivial to me. Chrome already doesn't adhere to the spec anyway by allowing `/*comments*/`. In JavaScript object keys are ordered, but it also doesn't matter because the only meaningful consumer of this section is the browser.

P.S. Preserving order of object keys when processing JSON is a common task with solutions in many languages, it's not something outrageous.

**Comment by xeenon** [comment]
Often JSON is turned into data structures internally that don't support order. That is the case in Safari, WebKit, Swift, Cocoa, etc. I don't support a custom JSON parser for this use case.

**Comment by tophf** [comment]
Here's another idea: put the browser-specific `background` into `browser_specific_settings` and let it override the main one.

**Comment by regseb** [comment]
The `"background"` property can take a background array. The browser tries to implement the first background listed. In the example below, Chromium, which supports service workers, uses the first background. Firefox uses the second. When Firefox supports service workers, it will use the first.

```json
{
  "manifest_version": 3,
  "background": [{
    "service_worker": "background/service_worker.js",
  }, {
    "scripts": ["background/script.js"],
  }]
}
```

For backwards compatibility, the `"background"` property also supports an object directly.

Here's a TypeScript definition of the `"background"` property:

```typescript
type BackgroundServiceWorker = {
  service_worker: string,
  type?: "classic" | "module",
};

type BackgroundScripts = {
  scripts: string[],
  type?: "classic" | "module",
};

type BackgroundPage = {
  page: string,
  persistent?: boolean,
  type?: "classic" | "module",
};

type BackgroundArray = (BackgroundServiceWorker | BackgroundScripts | BackgroundPage)[];

type Background =
  | BackgroundServiceWorker
  | BackgroundScripts
  | BackgroundPage
  | BackgroundArray;

**Comment by carlosjeurissen** [comment]
> @xeenon: Okay, what should the "preferred environment" key be? The original proposal was `preferred_environment` set to either `serviceWorker` or `page`. I tend to prefer `service_worker` over `serviceWorker` since it matches the keys at the top level.

> @erosman: AFA "preferred environment" key, why not use the same key as in "background"? That way, there wont be a need for mapping

The reasoning behind going for `serviceWorker` and `page` is to be independent on how the background page script has been defined (either via `page` or `scripts`). This would keep the window open for having a generated serviceWorker based on the `scripts` property. In the case `page` is preferred, the browser can use `page` directly if defined, else fall back to `scripts` and compose a `_generated_background_page.html` file. Using `service_worker` would assume you can also set it to `scripts`, which would be ambiguous.

> @tophf: Wouldn't it be simpler and intuitive for developers if browsers simply treat this section as ordered dictionary?

Also would not be supportive of this. Even tho sometimes the order is kept as is, many processors, tools, build scripts, browsers, actually assume the order does not matter. We can not rely on this.

> @regseb: The `"background"` property can take a background array. The browser tries to implement the first background listed. In the example below, Chromium, which supports service workers, uses the first background. Firefox uses the second. When Firefox supports service workers, it will use the first.

How is this easier than having a `preferred_environment` property stating a background page or a serviceWorker should be preferred?

> @erosman: If they want  "service_worker" for all browser, then they will only include  "service_worker"
> If they want  "service_worker" for Chrome and  "scripts" for others, then they will include both

The initial reasoning behind supporting both `service_worker` and either `scripts` or `page` is for backward compatibility. Going for just `service_worker` in your manifest would break all compatibility with all firefox builds prior to Firefox adding support of `service_worker` (which is not yet supported as of today).
This is why having a `preferred_environment` solution is able to suite the following scenarios:
1) browser supports the preferred environment, preferred environment will be used
2) browser only supports a serviceWorker, preferred is page, serviceWorker will be used (chrome)
3) browser only supports a page, preferred is serviceWorker, will use a page (older firefox and safari)
4) browser supports both serviceWorker & page, no preferred environment set. The preferred environment of the browser will be used (right now this seems to be serviceWorker for all browsers)

> @WorldLanguages: However, the following behavior from Firefox is making me very confused. Firefox considers `background.service_worker` an "unexpected property", but **it acts differently when encountered with it**, and for some reason the background script doesn't run! Changing it to `background.service_workerrrr` (or anything else) fixes the issue. What's going on?

Do you know in what Firefox version update it stopped loading `background.scripts` when `background.service_worker` is defined as well?

> @tophf: Here's another idea: put the browser-specific `background` into `browser_specific_settings` and let it override the main one.

This would make sense if one wants to prefer a serviceWorker in Firefox yet a page in Safari. Or the reverse. Is this scenario very likely? I think not.

**Comment by erosman** [comment]
> This is why having a `preferred_environment` solution is able to suite the following scenarios:
> 
>     1. browser supports the preferred environment, preferred environment will be used
> 
>     2. browser only supports a serviceWorker, preferred is page, serviceWorker will be used (chrome)
> 
>     3. browser only supports a page, preferred is serviceWorker, will use a page (older firefox and safari)

There is a missing scenario:

4.  browser that supports both  serviceWorker & page (scripts), (and not preferred environment yet)....


Please note that there are 3 options for the "background", plus a "type"

```json
"background": {
  "service_worker": "background.js",
  "scripts": ["script.js"],
  "page": "background.html"
  "type": "module"
}
```

Using `serviceWorker` and `page` can result in ambiguity where a `page` can refer to "page" property or "scripts" property.

Therefore ....

```
{
  "preferred_environment": "service_worker" or "scripts" or "page'
}
```

**Comment by carlosjeurissen** [comment]
> @erosman: There is a missing scenario:
> 4. browser that supports both  serviceWorker & page (scripts), (and not preferred environment yet)....

In that case the browser decides. Currently all browsers seem to prefer serviceWorker if both are defined. Which seems logical as in most situations, a serviceWorker is supposed to be more performant.

> @erosman: Please note that there are 3 options for the "background", plus a "type"
> 
> ```json
> "background": {
>   "service_worker": "background.js",
>   "scripts": ["script.js"],
>   "page": "background.html"
>   "type": "module"
> }
> ```
> 
> Using `serviceWorker` and `page` can result in ambiguity where a `page` can refer to "page" property or "scripts" property. Therefore ....
>   "preferred_environment": "service_worker" or "scripts" or "page'

As far as I know defining both `page` and `scripts` has always been invalid in the history of browser extensions. Even when including both would be valid, why would someone prefer `scripts` over `page`? It can simply not define `page` and use `scripts` only. This is why I believe setting `preferred_environment` to `page` should be enough for both the specified `background.page` and the `_generated_background_page.html` based on `background.scripts`.

This is why it is called `preferred_environment`. Scripts is not an environment in my opinion. `scripts` in the classic sense still uses a `_generated_background_page.html` and thus is a `page` environment.

**Comment by tophf** [comment]
> Currently all browsers seem to prefer serviceWorker if both are defined. Which seems logical as in most situations, a serviceWorker is supposed to be more performant.

No, it's not really logical because in most situations there's no difference in performance. There's only a few super rare cases where the extension is doing something super heavy on its main thread for extended periods of time. Generally, I'd prefer non-SW background script because it provides a standard JS environment with all features enabled, unlike the SW which breaks dozens of important features with the only workaround via the offscreen document at the cost of doubled CPU/memory consumption compared to a single non-SW background context.

**Comment by carlosjeurissen** [comment]
> > Currently all browsers seem to prefer serviceWorker if both are defined. Which seems logical as in most situations, a serviceWorker is supposed to be more performant.
> 
> No, it's not really logical because in most situations there's no difference in performance. There's only a few super rare cases where the extension is doing something super heavy on its main thread for extended periods of time. Generally, I'd prefer non-SW background script because it provides a standard JS environment with all features enabled, unlike the SW which breaks dozens of important features with the only workaround via the offscreen document at the cost of doubled CPU/memory consumption compared to a single non-SW background context.

Which is why I am advocating for introducing a `preferred_environment` property which allows you to prefer a page over a serviceWorker. This way you as developer can indicate this.

**Comment by WorldLanguages** [comment]
I understand that Firefox/Safari may need some more time to discuss a way to decide which environment to use, assuming both event pages and service workers are supported in the future.

However, I'm hopeful Chrome can convert the hard-errors to warnings soon. It's being tracked here: https://bugs.chromium.org/p/chromium/issues/detail?id=1418934

**Comment by oliverdunk** [comment]
I've just committed the change to move to warnings for the `scripts`, `page` and `persistent` keys in MV3 in Chrome: https://chromiumdash.appspot.com/commit/da4d041076acade2219c9bc4ad0c95bfd2f924ee. This should be available in Canary later today and in stable from Chrome 121.

Chrome also supports the `service_worker` key in MV2 extensions. We were considering a change to allow specifying both `scripts` and `service_worker` in that case and to prefer `service_worker`, but it seems like we should wait for the outcome of the discussions above.

**Comment by WorldLanguages** [comment]
Thanks oliverdunk for the Chrome update on its JSON validator.

> I understand that Firefox/Safari may need some more time to discuss a way to decide which environment to use, assuming both event pages and service workers are supported in the future.

Hopefully this will be resolved in stable versions a few months before [June 2024](https://developer.chrome.com/docs/extensions/migrating/mv2-sunset/#june-2024-chrome-mv2-deprecation-pre-stable-rollout), so that existing MV2 extensions can easily upgrade to MV3 while maintaining a single manifest file.
Failing to do so could cause unnecessary complexity for extension developers, and possibly fragmentation between extension stores. This is already the case with new extensions, but it's yet to be seen if it will affect extensions that have been maintained for the last 2 years or more. Fixing this issue would also simplify the creation of new cross-browser extensions as well.

**Comment by WorldLanguages** [comment]
By the way, it looks like MDN has a new section on scripts/page/service_worker browser support: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support

It is very useful! Thanks to the Firefox team for writing it!

**Comment by asamuzaK** [comment]
I think it would be great if we could combine `manifest.json` into one.
But the next question is: what do you expect to happen if unknown permissions are requested? For example, "offscreen" is not implemented in Firefox or Safari.

**Comment by carlosjeurissen** [comment]
@asamuzaK Ideally those permissions would be specified in the `optional_permissions`. And on requesting unsupported permissions the request to `browser.permissions.request` will reject. Just tested this in Chrome and seems this is already the case.

**Comment by asamuzaK** [comment]
> Ideally those permissions would be specified in the `optional_permissions`. 

Does that mean that even if it is a required permission, it should be put in `optional_permissions`?

**Comment by Juraj-Masiar** [comment]
I don't think that's a good idea. There is [many permissions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions#browser_compatibility) that don't have "optionality" implemented.

I think for future-proofing it's best to ignore unknown permissions (and even unknown manifest fields) and log a warning.

**Comment by WorldLanguages** [comment]
As far as I know, all browsers log warnings when they see unknown permissions. They don't throw errors. So it's already cross-browser compatible.

**Comment by regseb** [comment]
I've tried publishing [my extension](https://github.com/regseb/castkodi/blob/v7.5.1/src/manifest.json#L30-L34) on [Microsoft Edge Add-ons](https://microsoftedge.microsoft.com/), but I get these errors:

> The following checks failed:
>
> - Package acceptance validation error: **The background.scripts field cannot be used with manifest version 3. Use the background.service_worker field instead.** Error code: background: { "service_worker": "background.js", "scripts": [ "background.js" ], "type": "module" } Line: 30 Column: 17
> - Package acceptance validation error: The manifest shouldn't contain the key field.

---

Edit: @oliverdunk https://github.com/microsoft/MicrosoftEdge-Extensions/issues/136

**Comment by oliverdunk** [comment]
Thanks for flagging @regseb!

This sounds like a linter error that would need to be fixed on the Microsoft side. @mukul-p, is that something you might be able to pass on? I'm also aware of the https://github.com/microsoft/MicrosoftEdge-Extensions repository which could be a good place to file this.

**Comment by xeenon** [comment]
With https://github.com/WebKit/WebKit/pull/25678, WebKit (and future Safari releases) will now prefer `scripts` and `page` over `service_worker` if all three are supplied. We find the page based background pages to be better supported in our implementation currently.

**Comment by Rob--W** [comment]
An extension can specify `background.scripts` and `background.service_worker` and have it working in Firefox 121+ and Chrome 121+, which is also mentioned in https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/121#changes_for_add-on_developers

Because Chrome is not planning to support event pages in MV3 ever, I expect it to be likely for extension developers to target Chrome with `service_worker`. Although Firefox does not support `service_worker` yet, it might be at some point. If we were to prefer `service_worker` over `scripts` at that point, then existing Firefox extensions that use this `scripts` + `service_worker` trick would break. For long-term stability and reliability, we (in Firefox) are also inclined to prefer `scripts` (and `page`) over `service_worker` when multiple are specified.

**Comment by carlosjeurissen** [comment]
@xeenon @Rob--W Thanks for the updates! And very happy we got this far!

As for moving forward. Would Apple and Mozilla be open to offer developers to explicitly share their preferred environment using the proposed `preferred_environment`? This way we get a non-breaking long-term solution with maximum browser backward compatibility.

**Comment by xeenon** [comment]
@carlosjeurissen Yes, I would like to support `preferred_environment`.

**Comment by carlosjeurissen** [comment]
@xeenon That is wonderful. Do we agree on the naming and syntax?
The proposal so far has been to:
add a `preferred_environment` key under the `background` dictionary which can be set to either `serviceWorker` or `page`. If set to `page`, it would either use `scripts` or `page` depending on which one is specified. If `serviceWorker`, it will use the script defined in `service_worker`, while opening up a scenario in which we want to add some form of `_generated_service_worker.js` based on the `scripts` key.

The default order is up for browsers to decide. Currently it seems
Chrome will only use `serviceWorker` nomatter what
Firefox will prefer `serviceWorker` if defined and implemented
Safari will prefer `page` over `serviceWorker` when both are defined due to better support

Do we agree on this? I can proceed on making Bugzilla tickets for Mozilla and WebKit?

**Comment by xeenon** [comment]
@carlosjeurissen That looks good to me. I would just use `service_worker` for the value and not `serviceWorker`, for consistency.

Also what about `document` instead of `page` to avoid confusion since `page` is technically deprecated.

**Comment by carlosjeurissen** [comment]
@xeenon I like the use of `document` instead of `page`. Good one! As for `service_worker` seems based on the discussion in https://github.com/w3c/webextensions/issues/563, using snake case would not be out of place. Would work for me!

@Rob--W any thoughts?

**Comment by xeenon** [comment]
@carlosjeurissen I talked with @Rob--W about this some today, and he was suggesting this likely should take an array, for future compatibility and allow extension developers to specify their preferred fallback environments.

So `"preferred_environments": [ "document", "service_worker" ]` would prefer document, then service worker.

That then begs the question, should this be `supported_environments` or maybe just `environments`, where it is a defined list of what the developer expects (i.e. known working environments), and if the browser does not support anything in the list, it will fail to load the extension?

**Comment by WorldLanguages** [comment]
> if the browser does not support anything in the list, it will fail to load the extension

To clarify, this means that it would only fail if ALL the environments are unavailable, right?

**Comment by xeenon** [comment]
@WorldLanguages Yes

**Comment by carlosjeurissen** [comment]
@xeenon going for `supported_environments` or `environments` would be very ambiguous. Based on what keys the extension provides, it is already clear what `environments` the extension itself supports. The concept behind `preferred` is the fact browsers can decide to ignore this preference (Chromium)). Tho I agree `preferred_environments` may not be the best naming. We could think of alternatives like `preference: ["document", "service_worker"]`. Which drops the uncomfortable term `environments`.

Happy with it being an array. This indeed opens up potential future additions and makes it even more explicit which environments in what order one is dealing with.

**Comment by xeenon** [comment]
@carlosjeurissen Makes sense. I like `preference`. It does not limit the interpretation to a specific type of content or execution context. That said, another option could be `preferred_context`, or just `context` if preference is implied and can be ignored by the browser.

Ideally we can have it accept a string or array of strings, so I kind of walk back my preference for a plural term.

**Comment by xeenon** [comment]
@carlosjeurissen Another option I thought about is `precedence`.

```json
"precedence": ["document", "service_worker"]
```

**Comment by xeenon** [comment]
@carlosjeurissen Bikeshedding aside, what should we do in cases likes these?

```json
"background": {
    "preferred_environment": [ "service_worker" ],
    "service_worker": "background.js",
    "scripts": [ "background.js" ]
}
```

```json
"background": {
    "preferred_environment": [ "document" ],
    "page": "background.html",
    "scripts": [ "background.js" ]
}
```

Preferring specific keys like `page` or `service_worker` over `scripts` aligns better with Chrome’s choices and keeps our approach consistent. Even if `page` is deprecated, it matches well with `service_worker`, making our logic clearer and more predictable.

_I'm working on an implementation now for WebKit, so I'm trying to sort these things out._

**Comment by carlosjeurissen** [comment]
@xeenon as far as I know in Chrome and Firefox, specifying both "scripts" and "page" has always been invalid and rejected when attempting to load an extension.

As for your cases:
 ```json
"background": {
    "preferred_environment": [ "service_worker" ],
    "service_worker": "background.js",
    "scripts": [ "background.js" ]
}
```
In the above case, the `service_worker` will always be used. Except for browsers who do not understand `preferred_environment` and prefer `scripts` over `service_worker`.

```json
"background": {
    "preferred_environment": [ "document" ],
    "page": "background.html",
    "scripts": [ "background.js" ]
}
```
In the above case, browsers would not allow loading the extension because both `page` and `scripts` are specified. Which would make it invalid.

If we want to move away from this limitation and allow both to be specified. Then it would indeed make sense to use `page` and `scripts` as is. However, this would prevent a future implementation of `_generated_service_worker.js` based on the scripts specified in `scripts`, if we would want something like that.

To make sure we include each scenario. I assume we do not want to consider allowing multiple background instances to ever exist in the future right? For example running a serviceWorker side by side to a document or something similar.

**Comment by xeenon** [comment]
In Safari it is not an error to provide `page` and `scripts`. So we need to prefer one over the other.

**Comment by carlosjeurissen** [comment]
@xeenon About the naming.

`precedence` works for me. The downside of `preferred_environment` is the fact it implies singular, yet takes an array. While both `precedence` and `preference` lack this kind of singular / plural implication.

**Comment by carlosjeurissen** [comment]
> @xeenon Even if page is deprecated

Not under the impression `page` was ever deprecated. (except by chrome in mv3).

Glad to see the browser behaviour when including multiple values is now documented on mdn:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support

@xeenon Not sure the info here is correct. It states:

> mdn: supports background.scripts (or background.page) if service_worker is not specified.

This should probably be
```
Supports specifying scripts, page and service_worker simultaneously (since Safari x?).
```

> in Safari, the service_worker property is used, and a service worker starts that opens the tab because Safari gives priority to using service workers for background scripts.

This should probably be
```
Safari prefers `scripts`, then `page`, then `service_worker` since Safari x. Until Safari x, a `service_worker` is preferred over `page` and `scripts.
```

However these changes might not have shipped yet?

**Comment by xeenon** [comment]
@carlosjeurissen Correct, the recent changes have not shipped yet. The preference order change is in Safari Technology Preview 192, but that's all for now. We tend to update MDN once we ship things in a main Safari release.

**Comment by carlosjeurissen** [comment]
@oliverdunk When using this syntax on CWS, extensions currently seem to be rejected with the following message:
`The "background.page" key cannot be used with manifest_version 3. Use the "background.service_worker" key instead."`

The background property in the manifest is set as follows:
```json
"background": {
  "service_worker": "/background.js",
  "page": "/views/background.html",
  "preferred_environment": ["document", "service_worker"]
}
```

**Comment by oliverdunk** [comment]
Thanks @carlosjeurissen - I've escalated this internally, I'll let you know directly once I hear back. For the benefit of others, the item was rejected during review. Chrome fully supports this and the Chrome Web Store allows the extensions to be uploaded, but it looks like this was a one-off where it was incorrectly flagged. It shouldn't be a widespread problem :)

**Comment by carlosjeurissen** [comment]
@oliverdunk Thanks for escalating! If the CWS review team is afraid including both `service_worker` and `page` / `scripts` is not intentional, it could require the `preferred_environment`?

**Comment by oliverdunk** [comment]
> @oliverdunk Thanks for escalating! If the CWS review team is afraid including both `service_worker` and `page` / `scripts` is not intentional, it could require the `preferred_environment`?

I think it's more that your extension got tested using an older version of Chrome, so it actually failed to load in the browser. They are usually pretty good about using an updated version of Chrome, so I'm not sure what happened, but looking into it.

**Comment by dotproto** [comment]
Updated labels to reflect that Firefox and Chrome now support declaring event pages and a `service_worker` in the same manifest.json file.

<details name="2024-08-29-sample">
<summary><code>scripts</code> example</summary>

```json
"background": {
  "scripts": ["background.js"],
  "service_worker": "background.js"
}
```

</details>

<details name="2024-08-29-sample">
<summary><code>page</code> example</summary>

```json
"background": {
  "page": "background.html",
  "service_worker": "background.js"
}
```

</details>

**Comment by erosman** [comment]
Was a priority order established?

What would be loaded (now and/or in future when all are supported)  by Firefox, Chrome & Safari with the following entry?

```json
"background": {
  "scripts": ["background.js"],
  "service_worker": "background.js"
}
```

### preferred_environment

The important aim of dealing with inconsistencies in the `manifest.json` is to enable using a unified manifest for all browsers.
As `preferred_environment` would be different on each browsers, having to specify it in the `manifest.json` appears to defeat the purpose.

**For Chrome**

```json
"background": {
    "preferred_environment": [ "service_worker" ],
    "service_worker": "background.js",
    "scripts": [ "background.js" ]
}

```

**For Firefox**

```json
"background": {
    "preferred_environment": [ "document" ],
    "page": "background.html",
    "scripts": [ "background.js" ]
}
```


If using a `preferred_environment` is preferred, an alternative format might be more practical and can be used for all browsers.

```json
"background": {
    "preferred_environment": {
      "firefox": "scripts",
      "chrome": "service_worker",
      "safari": "scripts"
    },
    "service_worker": "background.js",
    "scripts": [ "background.js" ]
}

```

**Comment by carlosjeurissen** [comment]
@erosman with:
```json
"background": {
  "scripts": ["background.js"],
  "service_worker": "background.js"
}
```

As rob mentioned before: Per https://github.com/w3c/webextensions/issues/282#issuecomment-1998039421, Safari prefers scripts, page, then service_worker when they are all specified; Previously service_worker came first. When Firefox ships service_worker support, it will also prefer Event pages over service workers (https://github.com/w3c/webextensions/issues/282#issuecomment-1998053719).

Chrome will simply ignore the `scripts`, `page` and `preferred_environment`. Effectively preferring `service_worker`

@erosman Why would `preferred_environment` be different for each browser?

Use `document` to prefer a document background page (which can then be defined using either `scripts` or `page`.
Use `service_worker` to prefer a service_worker (defined using `service_worker` or, if supported (currently only safari), defined using `scripts`).

If however you want to prefer `document` in Firefox and Safari, but not in Chrome. You can still specify `document`, as Chrome will simply skip the `document` value (currently Chrome just fully skipps the `preferred_environment`, thus having the same result).

**Comment by erosman** [comment]
@carlosjeurissen I see... therefore, in future when both service `service_worker` & `scripts` are supported by Firefox & Safari:
 
 In the following example:

```json
"background": {
  "preferred_environment": [ "scripts" ],
  "service_worker": "worker.js",
  "scripts": [ "background.js" ]
}
```

- Chrome loads `service_worker`
- Firefox & Safari load `scripts`

What would be the outcome of the following?

```json
"background": {
  "service_worker": "worker.js",
  "scripts": [ "background.js" ]
}
```
 - Chrome loads `service_worker`
- Firefox & Safari ??


For the sake of being thorough, I would imagine  `preferred_environment` would be ignored in the following:

```json
"background": {
  "preferred_environment": [ "scripts" ],
  "service_worker": "worker.js"
}
```

**Comment by carlosjeurissen** [comment]
@erosman The outcome of the following would be:

```json
"background": {
  "service_worker": "worker.js",
  "scripts": [ "background.js" ]
}
```

Chrome loads: service_worker
Firefox & Safari: load scripts in a `_generated_background.html` "document" using `background.js`.

As for:
```json
"background": {
  "preferred_environment": [ "scripts" ],
  "service_worker": "worker.js"
}
```

For the extension developer, having the browser give a warning or even an error when an extension is loaded with this manifest would be preferred as `scripts` is not a valid value for `preferred_environment`. Even if `document` would have been specified a warning would be very welcome as no information is there in the background property to create a background document.

This is actually why I suggested to use `serviceWorker` instead of `service_worker` for preferred_environment to remove the perceived connection between the background keys and the environment.

**Comment by xeenon** [comment]
Indeed, `scripts` is not an environment, it is an input in this case. Safari [supports](https://github.com/WebKit/WebKit/commit/6d27093d068a866cc0f2948e561882d1f2d8039b) `document` and `service_worker` as valid `preferred_environment` values. We will also generate a `_generated_service_worker.js` script based on `scripts` and `type` keys when `preferred_environment` is `service_worker`.

@carlosjeurissen `serviceWorker` vs `service_worker` is too subtle of a distinction to be truly different. We might as well match the snake case at that point. I could buy `worker` being the environment name.

**Comment by dotproto** [comment]
Closing this issue as it has been implemented in all browsers.

**Comment by dotproto** [comment]
Reopening. I didn't realize that implementation of `preferred_environment` was keeping this issue open.

**Comment by carlosjeurissen** [comment]
In addition to the `preferred_environment` there is also the generated service worker based on both the `scripts` and `module` keys which Safari implemented.

**Comment by carlosjeurissen** [comment]
For Firefox, this is tracked here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1930334
https://phabricator.services.mozilla.com/D228538

**Comment by Rob--W** [comment]
Note: Safari 18 defaults to an event page ("scripts" or "page") over a service worker ("service_worker") instead of the opposite, if both are specified (see commit: https://commits.webkit.org/275916@main). The `preferred_environment` can be used to configure the preferred behavior (see commit: https://commits.webkit.org/277228@main).
This matches Firefox's behavior if it ever enables support for service workers (thanks to @carlosjeurissen, Firefox 136+ recognizes the `preferred_environment`key).

Safari 18 also introduces support for auto-generating the worker from the `"scripts"` array, if `preferred_environment` is present and specified to prioritize `service_worker`. Neither Firefox nor Chrome have implemented that.

The cross-browser documentation for this behavior is at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background#browser_support . Safari's changes are not included yet, that will be done as part of https://github.com/mdn/content/issues/35630

**Comment by erosman** [comment]
To clarify, is the following the agreed behaviour when BOTH  `service_worker` & `scripts/page` are present?

### Chrome
- loads `service_worker`
- Ignores others

### Firefox & Safari
- loads `preferred_environment`, if present
- Otherwise, loads event page ("scripts" or "page")

**Comment by Rob--W** [comment]
Yes, but only when Firefox enables support for service_worker, which it has not yet. Firefox only supports document-based background pages for now.

**Comment by erosman** [comment]
Bearing in mind the objective of  cross-browser `manifest.json` unification, if/once both `service_worker` & `scripts/page` are supported by Firefox & Safari, please note the following:



<table>
  <tr>
    <th>manifest.json</th>
    <th>Chrome</th>
    <th>Firefox & Safari</th>
  </tr>
  <tr>
    <td>
<pre lang="json">
"background": {
  "scripts": [ "background.js" ],
  "service_worker": "worker.js",
  "type": "module"
}
</pre>
    </td>
    <td>service_worker</td>
    <td>scripts</td>
  </tr>
  <tr>
    <td>
<pre lang="json"> 
"background": {
  "preferred_environment": [ "document" ],
  "scripts": [ "background.js" ],
  "service_worker": "worker.js",
  "type": "module"
}
</pre>
    </td>
    <td>service_worker</td>
    <td>scripts</td>
  </tr>
  <tr>
    <td>
<pre lang="json">
"background": {
  "preferred_environment": [ "service_worker" ],
  "scripts": [ "background.js" ],
  "service_worker": "worker.js",
  "type": "module"
}
</pre>
    </td>
    <td>service_worker</td>
    <td>service_worker</td>
  </tr>
  <tr>
    <td>
<pre lang="json">
"background": {
  "service_worker": "worker.js",
  "type": "module"
}
</pre>
    </td>
    <td>service_worker</td>
    <td>service_worker</td>
  </tr>
</table>

### Chrome
- loads `service_worker` under all circumstances

### Firefox & Safari
- Without `preferred_environment` loads  `scripts`
- with `preferred_environment` set to `document`, loads  `scripts` _(`preferred_environment` is irrelevant)_
- with `preferred_environment` set to `service_worker`, loads  `service_worker` _(`"scripts"` is irrelevant)_

Therefore, the following appears to cover the objective adequately.

<table>
  <tr>
    <th>manifest.json</th>
    <th>objective</th>
  </tr>
  <tr>
    <td>
<pre lang="json">
"background": {
  "scripts": [ "background.js" ],
  "service_worker": "worker.js",
  "type": "module"
}
</pre>
    </td>
    <td>if "scripts" is preferred for Firefox & Safari</td>
  </tr>
  <tr>
    <td>
<pre lang="json">
"background": {
  "service_worker": "worker.js",
  "type": "module"
}
</pre>
    </td>
    <td>if "service_worker" is preferred for Firefox & Safari</td>
  </tr>
</table>

---

_P.S. Chrome outputs an error in Developer mode `'background.scripts' requires manifest version of 2 or lower.`
I think a silent warning would be preferred._

![Image](https://github.com/user-attachments/assets/a47c7c61-e00a-4a3e-a466-0c0a4d682ca5)

**Comment by carlosjeurissen** [comment]
@erosman background.preferred_environment should be an array. This has been designed like this on purpose. See:
https://github.com/w3c/webextensions/issues/282#issuecomment-2037773022

In addition, `scripts` is not a valid value for `preferred_environment`. You can either specify `document` or `service_worker`.

**Comment by erosman** [comment]
@carlosjeurissen Thank you. I will update my post. However, the point remains unchanged.

**Comment by xeenon** [comment]
@erosman Without Chrome support, this property offers limited benefits. Currently, its primary purpose is to future-proof our API surface, ensuring compatibility with potential future choices. Safari and Firefox plan to support multiple environments, and this property facilitates that by allowing extensions to specify only `scripts`, generating a service worker or page (with or without modules).

**Comment by carlosjeurissen** [comment]
@erosman in addition, Firefox original plan was to default to `service_worker` if both `scripts` and `service_worker` and are present once implemented.

In addition, once Mozilla would actually start supporting `service_workers`, leaving out `scripts` would break older Firefox versions and Firefox forks not yet supporting `service_workers`. Thus `preferred_environment` provides backwards compatibility for those situations.

**Comment by carlosjeurissen** [comment]
Edit:
moved the empty background scripts declarations to a new issue: https://github.com/w3c/webextensions/issues/784

As this issue has mostly been defined by allowing `scripts` alongside `service_worker` and the implemented labels reflect this, I have created a new issue for the `_generated_service_worker.js` here: https://github.com/w3c/webextensions/issues/785. Thus allowing this issue to be marked as done and closed.

@dotproto Can we rename `topic: service worker` to `topic: background scripting` to be more generic and inclusive?

**Comment by carlosjeurissen** [comment]
@Rob--W seems the validator used by [web-ext ](https://github.com/mozilla/web-ext) for example does not consider a `background.page` alongside `background.service_worker` a valid combination. It still complaints it needs `background.scripts` even though `background.page` alongside `background.service_worker` should work as well.

**Comment by carlosjeurissen** [comment]
FYI. Safari currently does not fall back to the document environment if the service worker does not load due to lockdown restrictions. see https://bugs.webkit.org/show_bug.cgi?id=314194

Orion currently incorrectly favors `background.scripts` if loaded as chrome extension. It also does not respect `background.preferred_environment`. See:
https://orionfeedback.org/d/13760-incorrect-background-environment-preference-in-extensions

For both, see demo extension:
https://jeurissen.co/webext-demos/background-preferred-environment-service-worker

**Comment by Rob--W** [comment]
> [@Rob--W](https://github.com/Rob--W) seems the validator used by [web-ext ](https://github.com/mozilla/web-ext) for example does not consider a `background.page` alongside `background.service_worker` a valid combination.

We explicitly consider the combination valid. And with `preferred_environment` specified to explicitly signal the intent, the warning is dropped entirely (this was implemented in https://github.com/mozilla/addons-linter/pull/5899, included in [addons-linter@9.5.0](https://github.com/mozilla/addons-linter/releases/tag/9.5.0), shipped with [web-ex@9.3.0](https://github.com/mozilla/web-ext/releases/tag/9.3.0)).
